lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Set;
+
+import org.apache.lucene.index.Term;
+
+/**
+ * Allows recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public abstract class QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param query the leaf query
+   * @param term  the term the query will match on
+   */
+  public void consumesTerm(Query query, Term term) { }
+
+  /**
+   * Called by leaf queries that do not match on terms
+   * @param query the query
+   */
+  public void visitLeaf(Query query) { }
+
+  /**
+   * Pulls a visitor instance for visiting child clauses of a query
+   *
+   * The default implementation returns {@code this}, unless {@code occur} is equal
+   * to {@link BooleanClause.Occur#MUST_NOT} in which case it returns
+   * {@link #EMPTY_VISITOR}
+   *
+   * @param occur   the relationship between the parent and its children
+   * @param parent  the query visited
+   */
+  public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+    if (occur == BooleanClause.Occur.MUST_NOT) {
+      return EMPTY_VISITOR;
+    }
+    return this;
```

Comment:
I think I had missed above that this returns itself, which kind of answers my question around where the state is. I do find this a bit hard to follow, is `getSubVisitor` used only for short-circuiting based on the occur argument? Are there cases where we return something that differs from empty or this?

Reply:
Have a look in QueryVisitorTests to see an example of a more complicated visitor, which builds a tree of conjunction/disjunction nodes and then selects minimum sets of terms that must be present in a document for that doc to match.

Reply:
makes sense, so I take it that such usage will become production code in the future and not only test code. is the parent argument needed ?

Reply:
> so I take it that such usage will become production code in the future and not only test code

Yes - I have plans for luwak :) . I've added some test code to show how the `parent` argument can be used as well.